File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/MediaAdapter.java
package app.editors.manager.ui.adapters;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Point;
import android.graphics.PorterDuff;
import android.media.ThumbnailUtils;
import android.provider.MediaStore;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

import app.editors.manager.R;
import app.editors.manager.app.Api;
import app.editors.manager.app.App;
import app.editors.manager.managers.tools.AccountSqlTool;
import app.editors.manager.managers.tools.CacheTool;
import app.editors.manager.managers.tools.PreferenceTool;
import app.editors.manager.managers.utils.GlideUtils;
import app.editors.manager.mvp.models.explorer.File;
import app.editors.manager.ui.adapters.base.BaseAdapter;
import butterknife.BindView;
import butterknife.ButterKnife;
import lib.toolkit.base.managers.tools.GlideTool;
import lib.toolkit.base.managers.utils.ContentResolverUtils;
import lib.toolkit.base.managers.utils.StringUtils;
import lib.toolkit.base.managers.utils.UiUtils;

public class MediaAdapter extends BaseAdapter<File> {

    private static final int ALPHA_DELAY = 500;
    private static final float ALPHA_FROM = 0.0f;
    private static final float ALPHA_TO = 1.0f;

    @Inject
    Context mContext;

    @Inject
    PreferenceTool mPreferenceTool;

    @Inject
    GlideTool mGlideTool;

    @Inject
    CacheTool mCacheTool;

    @Inject
    AccountSqlTool mSqlTool;

    private int mCellSize;

    public MediaAdapter(final int cellSize) {
        App.getApp().getAppComponent().inject(this);
        mCellSize = cellSize;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int typeHolder) {
        switch (typeHolder) {
            case TYPE_ITEM_ONE: {
                final View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.list_media_image, viewGroup, false);
                return new ViewHolderImage(view);
            }

            case TYPE_ITEM_TWO: {
                final View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.list_media_video, viewGroup, false);
                return new ViewHolderVideo(view);
            }

            default:
                throw new RuntimeException("Unknown type is unacceptable: " + typeHolder);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        final File file = getItem(position);
        switch (getExtension(position)) {
            case IMAGE:
            case IMAGE_GIF: {
                final ViewHolderImage mViewHolder = (ViewHolderImage) viewHolder;
                mViewHolder.bind(file);
                break;
            }

            case VIDEO_SUPPORT: {
                final ViewHolderVideo mViewHolder = (ViewHolderVideo) viewHolder;
                mViewHolder.bind(file);

                break;
            }
        }
    }

    @Override
    public int getItemViewType(int position) {
        switch (getExtension(position)) {
            case IMAGE:
            case IMAGE_GIF:
                return TYPE_ITEM_ONE;
            case VIDEO_SUPPORT:
                return TYPE_ITEM_TWO;
            default:
                return TYPE_UNKNOWN;
        }
    }

    private StringUtils.Extension getExtension(int position) {
        final File file = getItem(position);
        final String ext = file.getFileExst();
        return StringUtils.getExtension(ext);
    }

    private Map<String, String> getHeaders() {
        Map<String, String> headers = new HashMap<>();
        headers.put(Api.HEADER_AUTHORIZATION, mPreferenceTool.getToken());
        headers.put(Api.HEADER_ACCEPT, Api.VALUE_ACCEPT);
        return headers;
    }

    public void setCellSize(int cellSize) {
        mCellSize = cellSize;
    }

    protected class ViewHolderImage extends RecyclerView.ViewHolder {

        @BindView(R.id.media_image_layout)
        FrameLayout mImageLayout;
        @BindView(R.id.media_image_view)
        ImageView mImageView;
        @BindView(R.id.media_image_progress)
        ProgressBar mProgressBar;

        ViewHolderImage(final View view) {
            super(view);
            ButterKnife.bind(this, view);
            UiUtils.setLayoutParams(mImageLayout, mCellSize, mCellSize);
            mImageView.setClickable(false);
            mProgressBar.getIndeterminateDrawable().setColorFilter(ContextCompat.getColor(mContext, R.color.colorAccent), PorterDuff.Mode.SRC_ATOP);
            view.setOnClickListener(v -> {
                if (mOnItemClickListener != null) {
                    mOnItemClickListener.onItemClick(v, getLayoutPosition());
                }
            });
        }

        public void bind(File file) {
            final RequestListener<Bitmap> requestListener = new RequestListener<Bitmap>() {
                @Override
                public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Bitmap> target, boolean isFirstResource) {
                    UiUtils.setLayoutParams(mImageView, ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
                    UiUtils.setImageTint(mImageView, R.drawable.ic_media_error, R.color.colorLightWhite);
                    showImage();
                    return true;
                }

                @Override
                public boolean onResourceReady(Bitmap resource, Object model, Target<Bitmap> target, DataSource dataSource, boolean isFirstResource) {
                    mImageView.setImageBitmap(resource);
                    showImage();
                    return true;
                }
            };

            if (mSqlTool.getAccountOnline() != null && mSqlTool.getAccountOnline().isWebDav()) {
                mGlideTool.load(mImageView, GlideUtils.getWebDavUrl(file.getId(), mSqlTool.getAccountOnline()),
                        false, new Point(mCellSize, mCellSize), requestListener);
            } else if (file.getId().equals("")) {
                mGlideTool.load(mImageView, file.getWebUrl(), false, new Point(mCellSize, mCellSize), requestListener);
            } else {
                mGlideTool.load(mImageView, GlideUtils.getCorrectLoad(file.getViewUrl(), mPreferenceTool),
                        false, new Point(mCellSize, mCellSize),
                        requestListener);
            }

        }

        private void showImage() {
            mProgressBar.setVisibility(View.GONE);
            mImageView.setVisibility(View.VISIBLE);
            mImageView.setAlpha(ALPHA_FROM);
            mImageView.animate().alpha(ALPHA_TO).setDuration(ALPHA_DELAY).start();
        }
    }

    protected class ViewHolderVideo extends RecyclerView.ViewHolder {

        @BindView(R.id.media_video_layout)
        FrameLayout mVideoLayout;
        @BindView(R.id.media_video_view)
        ImageView mVideoView;
        @BindView(R.id.view_icon_background_layout)
        FrameLayout mViewIconBackgroundLayout;
        @BindView(R.id.view_icon_background_image)
        ImageView mViewIconBackgroundImage;

        ViewHolderVideo(final View view) {
            super(view);
            ButterKnife.bind(this, view);
            UiUtils.setLayoutParams(mVideoLayout, mCellSize, mCellSize);
            mVideoView.setClickable(false);
            mViewIconBackgroundLayout.setClickable(false);
            view.setOnClickListener(v -> {
                if (mOnItemClickListener != null) {
                    mOnItemClickListener.onItemClick(v, getLayoutPosition());
                }
            });
        }

        public void bind(File file) {
            mViewIconBackgroundLayout.setBackgroundResource(R.drawable.drawable_media_background_video_play_light);
            UiUtils.setImageTint(mViewIconBackgroundImage, R.drawable.ic_media_play, R.color.colorPrimary);
            if (file.getId().equals("")) {
                mCacheTool.getBitmap(file.getWebUrl(), bitmap -> {
                    if (bitmap == null) {
                        Bitmap b = ThumbnailUtils.createVideoThumbnail(file.getWebUrl(), MediaStore.Images.Thumbnails.MINI_KIND);
                        mVideoView.setImageBitmap(b);
                        mCacheTool.addBitmap(file.getWebUrl(), b, null);
                    } else {
                        mVideoView.setImageBitmap(bitmap);
                    }
                    showVideo();
                });
            } else {
                mCacheTool.getBitmap(file.getViewUrl(), bitmap -> getFrame(bitmap, file.getViewUrl()));
            }
        }

        private void getFrame(Bitmap bitmap, String url) {
            if (bitmap == null) {
                // TODO fix queue
                ContentResolverUtils.getFrameFromWebVideoAsync(url, getHeaders(), new Point(mCellSize, mCellSize),
                        bitmapFrame -> {
                            if (bitmapFrame != null) {
                                mVideoView.setImageBitmap(bitmapFrame);
                                mCacheTool.addBitmap(url, bitmapFrame, null);
                                showVideo();
                            }
                        });
            } else {
                mVideoView.setImageBitmap(bitmap);
                showVideo();
            }
        }

        private void showVideo() {
            mViewIconBackgroundLayout.setBackgroundResource(R.drawable.drawable_media_background_video_play_grey);
            mVideoView.setVisibility(View.VISIBLE);
            mVideoView.setAlpha(ALPHA_FROM);
            mVideoView.animate().alpha(ALPHA_TO).setDuration(ALPHA_DELAY).start();
        }
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/MediaAdapter.kt
package app.editors.manager.ui.adapters

import android.accounts.Account
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Point
import android.media.ThumbnailUtils
import android.provider.MediaStore
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.ProgressBar
import androidx.recyclerview.widget.RecyclerView
import app.documents.core.network.common.contracts.ApiContract
import app.documents.core.network.manager.models.explorer.CloudFile
import app.documents.core.storage.account.AccountDao
import app.documents.core.storage.account.CloudAccount
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.managers.tools.CacheTool
import app.editors.manager.managers.utils.GlideUtils
import app.editors.manager.ui.adapters.base.BaseAdapter
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import lib.toolkit.base.managers.tools.GlideTool
import lib.toolkit.base.managers.utils.AccountUtils
import lib.toolkit.base.managers.utils.ContentResolverUtils
import lib.toolkit.base.managers.utils.StringUtils
import lib.toolkit.base.managers.utils.StringUtils.getExtension
import lib.toolkit.base.managers.utils.UiUtils
import lib.toolkit.base.managers.utils.UiUtils.setImageTint
import lib.toolkit.base.managers.utils.UiUtils.setLayoutParams
import javax.inject.Inject

class MediaAdapter(cellSize: Int, private val scope: CoroutineScope) : BaseAdapter<CloudFile?>() {

    companion object {
        private const val ALPHA_DELAY = 500
        private const val ALPHA_FROM = 0.0f
        private const val ALPHA_TO = 1.0f
    }


    @Inject
    lateinit var context: Context

    @Inject
    lateinit var glideTool: GlideTool

    @Inject
    lateinit var cacheTool: CacheTool

    @Inject
    lateinit var accountDao: AccountDao

    private var cellSize: Int

    init {
        App.getApp().appComponent.inject(this)
        this.cellSize = cellSize
    }

    private val token = runBlocking(Dispatchers.Default) {
        accountDao.getAccountOnline()?.let { account ->
            AccountUtils.getToken(context, Account(account.getAccountName(), context.getString(lib.toolkit.base.R.string.account_type)))
                ?.let { token ->
                    return@runBlocking token
                }
        } ?: run {
            return@runBlocking ""
        }
    }

    private val headers: Map<String, String> by lazy {
        mapOf(
            ApiContract.HEADER_AUTHORIZATION to token,
            ApiContract.HEADER_ACCEPT to ApiContract.VALUE_ACCEPT
        )
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, typeHolder: Int): RecyclerView.ViewHolder {
        return when (typeHolder) {
            TYPE_ITEM_ONE -> {
                val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.list_media_image, viewGroup, false)
                ViewHolderImage(view)
            }
            TYPE_ITEM_TWO -> {
                val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.list_media_video, viewGroup, false)
                ViewHolderVideo(view)
            }
            else -> throw RuntimeException("Unknown type is unacceptable: $typeHolder")
        }
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        val file = getItem(position)
        when (getExtension(position)) {
            StringUtils.Extension.IMAGE, StringUtils.Extension.IMAGE_GIF -> {
                val mViewHolder = viewHolder as ViewHolderImage
                mViewHolder.bind(file)
            }
            StringUtils.Extension.VIDEO_SUPPORT -> {
                val mViewHolder = viewHolder as ViewHolderVideo
                mViewHolder.bind(file)
            }
            else -> {
                // Stub
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return when (getExtension(position)) {
            StringUtils.Extension.IMAGE, StringUtils.Extension.IMAGE_GIF -> TYPE_ITEM_ONE
            StringUtils.Extension.VIDEO_SUPPORT -> TYPE_ITEM_TWO
            else -> TYPE_UNKNOWN
        }
    }

    private fun getExtension(position: Int): StringUtils.Extension {
        val file = getItem(position)
        val ext = file!!.fileExst
        return getExtension(ext)
    }

    fun setCellSize(cellSize: Int) {
        this.cellSize = cellSize
    }

    inner class ViewHolderImage internal constructor(val view: View) : RecyclerView.ViewHolder(view) {

        private val imageLayout: FrameLayout = view.findViewById(R.id.media_image_layout)
        private val imageView: ImageView = view.findViewById(R.id.media_image_view)
        private val progressBar: ProgressBar = view.findViewById(R.id.media_image_progress)

        private val requestListener: RequestListener<Bitmap> = object : RequestListener<Bitmap> {
            override fun onLoadFailed(
                e: GlideException?,
                model: Any?,
                target: Target<Bitmap?>,
                isFirstResource: Boolean
            ): Boolean {
                setLayoutParams(
                    imageView,
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
                setImageTint(imageView, R.drawable.ic_media_error, lib.toolkit.base.R.color.colorTextSecondary)
                showImage()
                return true
            }

            override fun onResourceReady(
                resource: Bitmap,
                model: Any,
                target: Target<Bitmap?>,
                dataSource: DataSource,
                isFirstResource: Boolean
            ): Boolean {
                imageView.setImageBitmap(resource)
                showImage()
                return true
            }
        }

        init {
            setLayoutParams(imageLayout, cellSize, cellSize)

            imageView.isClickable = false
            UiUtils.setColorFilter(progressBar.context, progressBar.indeterminateDrawable, lib.toolkit.base.R.color.colorSecondary)

            view.setOnClickListener { v: View? ->
                mOnItemClickListener?.onItemClick(v, layoutPosition)
            }
        }

        fun bind(file: CloudFile?) {
            scope.launch {
                accountDao.getAccountOnline()?.let { account ->
                    when {
                        account.isWebDav && file?.id != "" -> {
                            loadWebDav(file, account)
                        }
                        file?.id == "" -> {
                            loadLocal(file)
                        }
                        else -> {
                            AccountUtils.getToken(
                                view.context,
                                Account(account.getAccountName(), view.context.getString(lib.toolkit.base.R.string.account_type))
                            )?.let { token ->
                                loadCloud(file, token)
                            }
                        }
                    }
                } ?: run {
                    if (file?.id == "") {
                        loadLocal(file)
                    }
                }
            }

        }

        private suspend fun loadCloud(file: CloudFile?, token: String) {
            val url = GlideUtils.getCorrectLoad(file?.viewUrl ?: "", token)
            withContext(Dispatchers.Main) {
                glideTool.load(imageView, url, false, Point(cellSize, cellSize), requestListener)
            }
        }

        private suspend fun loadLocal(file: CloudFile) {
            withContext(Dispatchers.Main) {
                glideTool.load(imageView, file.webUrl, false, Point(cellSize, cellSize), requestListener)
            }
        }

        private suspend fun loadWebDav(file: CloudFile?, account: CloudAccount) {
            AccountUtils.getPassword(
                view.context,
                Account(account.getAccountName(), view.context.getString(lib.toolkit.base.R.string.account_type))
            )?.let { pass ->
                val url = GlideUtils.getWebDavUrl(file?.id!!, account, pass)
                withContext(Dispatchers.Main) {
                    glideTool.load(imageView, url, false, Point(cellSize, cellSize), requestListener)
                }
            }
        }

        private fun showImage() {
            progressBar.visibility = View.GONE
            imageView.visibility = View.VISIBLE
            imageView.alpha = ALPHA_FROM
            imageView.animate().alpha(ALPHA_TO).setDuration(ALPHA_DELAY.toLong()).start()
        }
    }

    inner class ViewHolderVideo internal constructor(view: View) : RecyclerView.ViewHolder(view) {

        private val videoLayout: FrameLayout = view.findViewById(R.id.media_video_layout)
        private val videoView: ImageView= view.findViewById(R.id.media_video_view)

        init {
            setLayoutParams(videoLayout, cellSize, cellSize)
            videoView.isClickable = false
            view.setOnClickListener { v: View? ->
                mOnItemClickListener?.onItemClick(v, layoutPosition)
            }
        }

        fun bind(file: CloudFile?) {
            if (file!!.id == "") {
                cacheTool.getBitmap(file.webUrl) { bitmap: Bitmap? ->
                    if (bitmap == null) {
                        val b = ThumbnailUtils.createVideoThumbnail(file.webUrl, MediaStore.Images.Thumbnails.MINI_KIND)
                        videoView.setImageBitmap(b)
                        cacheTool.addBitmap(file.webUrl, checkNotNull(b), null)
                    } else {
                        videoView.setImageBitmap(bitmap)
                    }
                    showVideo()
                }
            } else {
                cacheTool.getBitmap(file.viewUrl) { bitmap: Bitmap? -> getFrame(bitmap, file.viewUrl) }
            }
        }

        private fun getFrame(bitmap: Bitmap?, url: String) {
            if (bitmap == null) {
                // TODO fix queue
                ContentResolverUtils.getFrameFromWebVideoAsync(url, headers, Point(cellSize, cellSize),
                    object : ContentResolverUtils.OnWebVideoListener {
                        override fun onGetFrame(bitmap: Bitmap) {
                            videoView.setImageBitmap(bitmap)
                            cacheTool.addBitmap(url, bitmap, null)
                            showVideo()
                        }

                    }, scope)
            } else {
                videoView.setImageBitmap(bitmap)
                showVideo()
            }
        }

        private fun showVideo() {
            videoView.visibility = View.VISIBLE
            videoView.alpha = ALPHA_FROM
            videoView.animate()?.alpha(ALPHA_TO)?.setDuration(ALPHA_DELAY.toLong())?.start()
        }

    }

}